pkts-sip/src/main/java/io/pkts/packet/sip/impl/SipUserHostParser.java

```diff
@@ -0,0 +1,353 @@
+package io.pkts.packet.sip.impl;
+
+import io.pkts.buffer.Buffer;
+import io.pkts.packet.sip.SipParseException;
+
+import java.io.IOException;
+
+import static io.pkts.packet.sip.impl.SipParser.*;
+
+/**
+ * Parses the user and host portion of the SIP URI.
+ *
+ * The reason why this method does both is because the userinfo portion is
+ * optional and as such you actually don't know whether the stuff you are
+ * currently going over is the hostport or the userinfo. Also, unfortunately
+ * the userinfo is allowed to have characters that normally are reserved
+ * (such as ';'), which complicates things as well.
+ *
+ */
+public class SipUserHostParser {
+    private final Buffer buffer;
+    private Buffer user;
+    private Buffer host;
+    private Buffer port;
+
+    // Host/Port parser
+    private enum HostPortParseState {
+        HOST, PORT, INVALID, END, OUT_OF_CHARS
+    }
+    private HostPortParseState parseState;
+    private int stateStartIndex;
+    private int stateCount;
+    private int hostPortEndIndex;
+
+    // Hostname parser
+    private enum HostnameParseState {
+        BEGIN, INVALID, IN_LABEL, IN_LABEL_DASH, AT_SEPARATOR
+    }
+    private HostnameParseState hostnameParseState;
+    private boolean hostnameValidTLD;
+
+    // IPv4 parser
+    private enum IPv4ParseState {
+        BEGIN, INVALID, ACCUMULATE_DIGITS, AT_SEPARATOR
+    }
+    private IPv4ParseState ipv4ParseState;
+    private int ipv4NumDigits;
+    private int ipv4NumSegments;
+
+    public SipUserHostParser(Buffer buffer) throws IOException {
```

Comment:
nit; final

and also typically you don't end up parsing from within the constructor. I prefer that you had a static parse method on the class and the class would then not be called a parser anymore but rather like so:

SipUserHostInfo.frame(final Buffer buffer)

Compare with how all other headers, addresses etc within pkts.io is done. E.g. SipURI.frame etc. Also, since you do not return a parser, the method to reset it doesn't make sense anymore.